### PR TITLE
JAMES-3265 Fix CassandraMessageMapper::deleteMessages mailbox counter…

### PR DESCRIPTION
**Pull request by chibenwa**
…s handling

This code path side effect was untested

### CODE REVIEW

File: mailbox/store/src/test/java/org/apache/james/mailbox/store/mail/model/MessageMapperTest.java [hunk 1]
```diff
@@ -377,6 +378,23 @@ void deleteShouldNotUpdateRecentWhenNotNeeded() throws MailboxException {
...
+        messageMapper.updateFlags(benwaInboxMailbox, new FlagsUpdateCalculator(new Flags(Flag.SEEN), FlagsUpdateMode.REPLACE), MessageRange.one(message3.getUid()));
+        messageMapper.updateFlags(benwaInboxMailbox, new FlagsUpdateCalculator(new Flags(Flag.SEEN), FlagsUpdateMode.REPLACE), MessageRange.one(message4.getUid()));
+
+        messageMapper.deleteMessages(benwaInboxMailbox, ImmutableList.of(message1.getUid(), message2.getUid(), message3.getUid()));
+
+        assertThat(messageMapper.getMailboxCounters(benwaInboxMailbox))
+            .isEqualTo(MailboxCounters.builder()
+                .mailboxId(benwaInboxMailbox.getMailboxId())
+                .count(2)
+                .unseen(1)
```
Comment (rouazana) [inline mailbox/store/src/test/java/org/apache/james/mailbox/store/mail/model/MessageMapperTest.java:394]: sorry, this class is hard to read I don't get why you expect 1 unseen at the end

Comment (chibenwa) [reply mailbox/store/src/test/java/org/apache/james/mailbox/store/mail/model/MessageMapperTest.java:394]: Sure. 

We have 5 messages:
 - **message1**, **message2**, **message3**, are deleted
 -  **message4** is seen
 -  **message4** is unseen (created in saveMessages)

Hence the expected count. Note that this test runs on the memory version not relying on denormalisation.

File: mailbox/store/src/test/java/org/apache/james/mailbox/store/mail/model/MessageMapperTest.java [hunk 2]
```diff
@@ -377,6 +378,23 @@ void deleteShouldNotUpdateRecentWhenNotNeeded() throws MailboxException {
         assertThat(messageMapper.findRecentMessageUidsInMailbox(benwaInboxMailbox)).containsOnly(message2.getUid(), message4.getUid());
     }
 
+    @Test
+    void deleteMessagesShouldCorrectlyUpdateMailboxCounters() throws MailboxException {
+        saveMessages();
+        messageMapper.updateFlags(benwaInboxMailbox, new FlagsUpdateCalculator(new Flags(Flag.SEEN), FlagsUpdateMode.REPLACE), MessageRange.one(message2.getUid()));
```
Comment (mbaechler) [inline mailbox/store/src/test/java/org/apache/james/mailbox/store/mail/model/MessageMapperTest.java:384]: could we write a decent api for changing a flag?

File: mailbox/store/src/test/java/org/apache/james/mailbox/store/mail/model/MessageMapperTest.java (hunk 1 above)
Comment (mbaechler) [reply mailbox/store/src/test/java/org/apache/james/mailbox/store/mail/model/MessageMapperTest.java:394]: we should rename the test.
Let me try: "deleteMessagesShouldDecrementUnseenToOneWhenDeletingOneUnseenMessageOnTwo"
I think the test would be easier to understand without message 1, 2 and 3.

File: mailbox/store/src/test/java/org/apache/james/mailbox/store/mail/model/MessageMapperTest.java (hunk 2 above)
Comment (chibenwa) [reply mailbox/store/src/test/java/org/apache/james/mailbox/store/mail/model/MessageMapperTest.java:384]: That's orthogonal to the bug reolution but let's add a default method then

Comment (chibenwa) [reply mailbox/store/src/test/java/org/apache/james/mailbox/store/mail/model/MessageMapperTest.java:384]: I wrote a :

```
    default Optional<UpdatedFlags> updateFlags(Mailbox mailbox, MessageUid uid, FlagsUpdateCalculator flagsUpdateCalculator) throws MailboxException {
        return Iterators.toStream(updateFlags(mailbox, flagsUpdateCalculator, MessageRange.one(uid)))
            .findFirst();
    }
```

Comment (mbaechler) [reply mailbox/store/src/test/java/org/apache/james/mailbox/store/mail/model/MessageMapperTest.java]: I was thinking about:
```
messageMapper.updateFlags(benwaInboxMailbox, message2.getUid(), flags -> flags.set(Flag.SEEN));
```

Comment (chibenwa) [reply mailbox/store/src/test/java/org/apache/james/mailbox/store/mail/model/MessageMapperTest.java]: That's not an easy change to unify `flags -> flags.set(Flag.SEEN)` into `FlagsUpdateCalculator`

Don't you mind if I don't invest for now in that refactoring?

Comment (mbaechler) [reply mailbox/store/src/test/java/org/apache/james/mailbox/store/mail/model/MessageMapperTest.java]: no problem

Comment (mbaechler) [reply mailbox/store/src/test/java/org/apache/james/mailbox/store/mail/model/MessageMapperTest.java]: look at https://github.com/linagora/james-project/pull/3537 and pick it if you want

Comment (chibenwa) [reply mailbox/store/src/test/java/org/apache/james/mailbox/store/mail/model/MessageMapperTest.java]: I will but in a follow up PR

**Comment by chibenwa** [comment]
Merged

**Pull request closed without merging**